src/shared/varlink.c

```diff
@@ -131,7 +131,9 @@ struct Varlink {
         JsonVariant *reply;
 
         struct ucred ucred;
+        struct ucred ucred_effective;
```

Comment:
as far as can see you use this structure for two purposes now: for impersonation when sending, and for storing the peer's ucred when receiving. And since varlink does both (i.e. send *and* receive) this is a feedback loop ;-)

Please split this up, and give this clear name, to minimize confusion. And maybe rename the pre-existing ucred structure too, to make this clear. maybe:

```c
struct ucred ucred_peer, ucred_send, ucred_recv;
bool ucred_peer_acquired:1, ucred_send_set:1, ucred_recv_acquired:1;
```

I think with that it should be clear what is what: ucred_peer → SO_PEERCRED data, i.e. always the connection time identity of the peer at time of connect(). ucred_send → SCM_CREDENTIALS data to send for impersonation purposes. ucred_recv → most recently read SCM_CREDENTIALS data received.